spec/src/main/asciidoc/core/invokers.asciidoc

```diff
@@ -0,0 +1,181 @@
+[[method_invokers]]
+== Method invokers
+
+CDI-based frameworks often need to invoke application methods declared on managed beans.
+Frameworks cannot invoke application methods directly, because they are not compiled against the application code.
+However, during application deployment, frameworks may observe application methods through CDI extensions and build an `Invoker` for each relevant method.
+The invokers can then be used at application runtime to invoke the methods indirectly.
+
+Method invokers are not supposed to be used by application code, as applications may invoke their own methods directly.
+
+[[building_invoker]]
+=== Building an `Invoker`
+
+The CDI container allows building an `Invoker` for a method of an <<enablement,enabled>> <<managed_beans,managed bean>>.
+The method for which the invoker is built is called the _target method_ of the invoker, and the managed bean is called the _target bean_ of the invoker.
+
+Invalid target methods are:
+
+* private methods,
+* constructors,
+* methods declared on the `java.lang.Object` class, except of the `toString()` method,
+* methods that are not declared on the bean class of the target bean or inherited from its supertypes.
+
+Attempting to build an invoker for an invalid target method leads to a deployment problem.
+
+When the target bean is not a managed bean, attempting to build an invoker leads to a deployment problem.
+When the target bean is an interceptor, attempting to build an invoker leads to a deployment problem.
+
+Multiple managed beans may inherit a method from a common supertype.
+In that case, an invoker must be built for each target bean individually.
+An invoker built for one target bean may not be used to invoke the target method on an instance of another target bean.
+
+The only way to build an invoker is using the `InvokerBuilder`.
+An `InvokerBuilder` can only be obtained in CDI portable extensions and build compatible extensions.
+See <<invoker_builder>> for more information.
+
+[[using_invoker]]
+=== Using an `Invoker`
+
+The `Invoker` interface contains a single method:
+
+[source,java]
+----
+public interface Invoker<T, R> {
+    R invoke(T instance, Object[] arguments);
+}
+----
+
+This section describes the behavior of the `invoke()` method.
+
+Calling `invoke()` invokes the target method on given `instance` of the target bean, passing given `arguments`.
+
+If the target method is `static`, the `instance` is ignored; by convention, it should be `null`.
```

Comment:
The below list (all of the `if` statements) might be better readable as bullet points.
That being said, all of this creates a hard to read section if you are just trying to read up on how to use invokers.
Personally I'd rather keep these details in the javadoc on the assumption that javadoc text is also considered a source of truth for how CDI should behave (which I am not sure it is). But that might be just me :)

Reply:
Also note that this would be even more verbose if we want to mention transformations and wrappers.

Reply:
Javadoc is very much a source of truth, which you can see in AtInject (which doesn't have anything else but javadoc). We also did this in Build Compatible Extensions, which are specified in full by the javadoc. Here, I wanted to take the exact opposite approach: put the gory details into the specification text, and leave the javadoc accessible to mere mortals.

Reply:
Transformers would be in a separate section and we would have to say that the rules specified for `Invoker.invoke()` apply after all transformations and lookups are complete. That should be enough.